### PR TITLE
1. fix(perf): Run CPU-intensive state updates in parallel rayon threads

### LINKED ISSUE #4790: Move note commitment tree updates to a dedicated blocking and CPU-heavy thread
## Motivation

Zebra's sync can be slow or fail, because note commitment tree updates involve some CPU-heavy cryptography.

So we want to move note commitment tree updates into a `rayon` CPU thread pool `scope`.

### API Reference

Blocking threads:
https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html

CPU-heavy threads:
https://docs.rs/rayon/latest/rayon/fn.scope_fifo.html

### Detailed Analysis

See https://github.com/ZcashFoundation/zebra/issues/4583#issuecomment-1186662862

### Designs

- [ ] Wrap the entire function that calls `append()`/`root()` in `rayon`, to minimise the cost of scheduling `tokio` and `rayon` threads
- [ ] We'd like to update sprout, sapling, and orchard note commitment trees in parallel, using Rayon's `join()`
- [ ] If there isn't a `tokio::spawn_blocking()` wrapping the update future, add it (see ticket #4779 and its PR)

Example code for `rayon` in `tokio` blocking threads:
https://github.com/ZcashFoundation/zebra/blob/81727d7df0ee57e6d4f2172c287e52e414c4e6a0/zebra-consensus/src/primitives/ed25519.rs#L121-L134

## Related Work

We'll move state updates to blocking tokio threads as part of #4779

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to improve Zebra's sync speed to avoid test timeouts in CI (#4155).

This PR moves CPU-intensive verification in state updates to the `rayon` thread pool.

Close #4790

Partially addresses #4583

Depends-On: #4199
Depends-On: #4795

### Specifications

https://docs.rs/rayon/latest/rayon/fn.in_place_scope_fifo.html

## Solution

- Run CPU-intensive chain validation and updates in parallel `rayon` threads
    - Run non-finalized validation and `Chain` updates in parallel
    - Run non-finalized sprout, sapling, and orchard tree updates and roots in parallel
    - After a non-finalized chain fork, do note commitment and history tree rebuilds in parallel
    - Run finalized sprout, sapling, and orchard tree updates and roots in parallel

Other performance improvements:
- Wrap `HistoryTree` in an `Arc` in the state 
- Skip generating unused interstitial Sprout treestates

Related changes:
- Remove redundant checks for empty shielded data

Test fixes:
- Improve anchor validation debugging and error messages
- Work around a test data bug, and save some CPU during tests

## Review

Anyone can review this PR, it's urgent because CI is failing, and it blocks a release.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

- #4788

### CODE REVIEW

File: zebra-state/src/service/check/anchors.rs [hunk 1]
```diff
@@ -159,10 +97,231 @@ pub(crate) fn anchors_refer_to_earlier_treestates(
...
+
+    Ok(())
+}
+
+/// This function fetches and returns the Sprout final treestates from the state,
+/// so [`sprout_anchors_refer_to_interstitial_treestates()`] can check Sprout
+/// final and interstitial treestates without accessing the disk.
+///
+/// Sprout anchors may also refer to the interstitial output treestate of any prior
+/// `JoinSplit` _within the same transaction_.
```
Comment (conradoplg) [inline zebra-state/src/service/check/anchors.rs]: I was a bit confused about these two functions but I get it now, here's a suggestion to hopefully make it clearer, feel free to modify

```suggestion
/// Sprout anchors may also refer to the interstitial output treestate of any prior
/// `JoinSplit` _within the same transaction_; these are created on the fly
/// in [`sprout_anchors_refer_to_treestates`].
```

File: zebra-state/src/service/check/anchors.rs [hunk 2]
```diff
@@ -159,10 +97,231 @@ pub(crate) fn anchors_refer_to_earlier_treestates(
...
+}
+
+/// Checks the Sprout anchors specified by transactions in `block`.
+///
+/// Sprout anchors may refer to some earlier block's final treestate (like
+/// Sapling and Orchard do exclusively) _or_ to the interstitial output
+/// treestate of any prior `JoinSplit` _within the same transaction_.
+///
+/// This method checks for anchors computed from the final and interstitial treestates of
+/// each transaction in the `prepared` block, using the supplied `sprout_final_treestates`.
```
Comment (conradoplg) [inline zebra-state/src/service/check/anchors.rs]: Same as above, feel free to modify

```suggestion
/// This method searches for anchors in the supplied `sprout_final_treestates` (which must be populated
/// with all treestates pointed to in the `prepared` block; see [`fetch_sprout_final_treestates`]) or in the interstitial
/// treestates which are computed on the fly in this function.
```

**Review by conradoplg — DISMISSED** [review]
Looks good, I was a bit confused with a refactoring and I added some documentation suggestions to hopefully make it clearer

**Comment by teor2345** [comment]
Sorry, I forgot to push the commits for:
    - Run non-finalized sprout, sapling, and orchard tree updates and roots in parallel
    - After a non-finalized chain fork, do note commitment and history tree rebuilds in parallel
    - Some minor cleanups

File: zebra-state/src/service/check/anchors.rs (hunk 1 above)
Comment (teor2345) [reply zebra-state/src/service/check/anchors.rs]: Applied in eebf3f581b

File: zebra-state/src/service/check/anchors.rs (hunk 2 above)
Comment (teor2345) [reply zebra-state/src/service/check/anchors.rs]: Applied in eebf3f581b

**Comment by teor2345** [comment]
I fixed some doc issues and added your suggestions, this is ready for another review.

**Review by conradoplg — APPROVED** [review]

**Pull request merged**